src/game/client/water/w3dwatertracks.cpp

```diff
@@ -85,35 +392,333 @@ void WaterTracksRenderSystem::Release_Resources()
+            }
+
+            obj->m_prevSystem = nullptr;
+            obj->m_nextSystem = m_freeModules;
+
+            if (m_freeModules) {
+                m_freeModules->m_prevSystem = obj;
+            }
+
+            m_freeModules = obj;
+        }
+    }
 }
 
 void WaterTracksRenderSystem::Reset()
 {
-#ifdef GAME_DLL
-    Call_Method<void, WaterTracksRenderSystem>(PICK_ADDRESS(0x007A18A0, 0x0064D221), this);
-#endif
+    WaterTracksObj *next;
+
+    for (WaterTracksObj *i = m_usedModules; i; i = next) {
+        next = i->m_nextSystem;
+        i->m_bound = false;
+        Release_Track(i);
+    }
 }
 
 void WaterTracksRenderSystem::Shutdown()
 {
-#ifdef GAME_DLL
-    Call_Method<void, WaterTracksRenderSystem>(PICK_ADDRESS(0x007A1930, 0x0064D264), this);
-#endif
+    WaterTracksObj *next;
+
+    for (WaterTracksObj *i = m_usedModules; i; i = next) {
+        next = i->m_nextSystem;
+
+        if (!i->m_bound) {
+            Release_Track(i);
+        }
+    }
+
+    while (m_freeModules) {
+        next = m_freeModules->m_nextSystem;
+        delete m_freeModules;
+        m_freeModules = next;
+    }
+
+    Ref_Ptr_Release(m_indexBuffer);
+    Ref_Ptr_Release(m_vertexMaterialClass);
+    Ref_Ptr_Release(m_vertexBuffer);
+}
+
+void WaterTracksRenderSystem::Update()
+{
+    static unsigned int iLastTime = rts::Get_Time();
+    iLastTime = rts::Get_Time();
+    WaterTracksObj *obj = m_usedModules;
+
+    while (obj) {
+        WaterTracksObj *next = obj->m_nextSystem;
+
+        if (!obj->m_bound || !obj->Update(33) && !obj->m_bound) {
+            Release_Track(obj);
+        }
+
+        obj = next;
+    }
 }
 
 void WaterTracksRenderSystem::Flush(RenderInfoClass &rinfo)
 {
-#ifdef GAME_DLL
-    Call_Method<void, WaterTracksRenderSystem, RenderInfoClass &>(PICK_ADDRESS(0x007A1A40, 0x0064D40B), this, rinfo);
+#ifdef BUILD_WITH_D3D8
+    if (g_theWriteableGlobalData->m_showSoftWaterEdge && g_theWaterTransparency->m_transparentWaterDepth != 0.0f) {
+        if (g_theWriteableGlobalData->m_useWaveEditor) {
+            Test_Water_Update();
+        }
+
+        Update();
+        rinfo.m_camera.Apply();
+
+        if (m_usedModules) {
+            if (!ShaderClass::Is_Backface_Culling_Inverted()) {
+                m_batchStart = 0xFFFF;
+                float r = g_theWriteableGlobalData->m_terrainAmbient[0].red;
+                float g = g_theWriteableGlobalData->m_terrainAmbient[0].green;
+                float b = g_theWriteableGlobalData->m_terrainAmbient[0].blue;
+                r = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + r;
+                g = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + g;
+                b = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + b;
+                r = r * 255.0f;
+                g = g * 255.0f;
+                b = b * 255.0f;
+                int color = (GameMath::Fast_To_Int_Truncate(r) << 16) | (GameMath::Fast_To_Int_Truncate(g) << 8)

```

Comment:
color isn't used? if so add a note to investigate this

Reply:
fixed.